url/url.cpp

```diff
@@ -191,6 +191,74 @@ std::string blob_url_create(Origin const &origin) {
     return result;
 }
 
+// https://url.spec.whatwg.org/#concept-host-serializer
+std::string Host::serialize() const {
+    if (type == HostType::Ip4Addr) {
+        return util::ipv4_serialize(std::get<std::uint32_t>(data));
+    } else if (type == HostType::Ip6Addr) {
+        return util::ipv6_serialize(const_cast<std::array<std::uint16_t, 8> &>(std::get<2>(data)));
+    }
+
+    return std::get<std::string>(data);
+}
+
+// https://url.spec.whatwg.org/#url-path-serializer
+std::string Url::serialize_path() const {
+    if (std::holds_alternative<std::string>(path)) {
+        return std::get<0>(path);
+    }
+
+    std::string output = "";
+
+    for (std::size_t i = 0; i < std::get<1>(path).size(); i++) {
+        output += "/" + std::get<1>(path)[i];
+    }
+
+    return output;
+}
+
+// https://url.spec.whatwg.org/#concept-url-serializer
+std::string Url::serialize(bool exclude_fragment) const {
+    std::string output = scheme + ":";
+
+    if (host.has_value()) {
+        output += "//";
+
+        if (!user.empty() || !passwd.empty()) {
```

Comment:
Call `includes_credentials` instead. It does the same thing and reads more like the spec.